sdk/storage/Azure.Storage.DataMovement/src/DownloadChunkHandler.cs

```diff
@@ -121,75 +142,125 @@ public struct Behaviors
             _downloadChunkEventHandler += DownloadChunkEvent;
         }
 
-        public void Dispose()
+        public async ValueTask DisposeAsync()
         {
-            CleanUp();
+            _downloadRangeChannel.Writer.Complete();
+            await _downloadRangeChannel.Reader.Completion.ConfigureAwait(false);
+            if (!_channelCancellationSource.IsCancellationRequested)
+            {
+                _channelCancellationSource.Cancel();
+            }
+            _channelCancellationSource.Dispose();
+
+            if (_currentBytesSemaphore != default)
+            {
+                _currentBytesSemaphore.Dispose();
+            }
+
+            DisposeHandlers();
         }
 
-        public void CleanUp()
+        public void DisposeHandlers()
         {
             _downloadChunkEventHandler -= DownloadChunkEvent;
         }
 
         private async Task DownloadChunkEvent(DownloadRangeEventArgs args)
         {
-            if (args.Success && !_cancellationToken.IsCancellationRequested)
+            if (args.Success)
+            {
+                await _downloadRangeChannel.Writer.WriteAsync(args, _cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // Report back failed event.
+                await InvokeFailedEvent(new Exception("Unexpected error: Experienced failed download range argument. " +
+                    $"Range: {args.Offset} - {args.BytesTransferred} with Transfer ID: {args.TransferId}")).ConfigureAwait(false);
+            }
+        }
+
+        private async Task NotifyOfPendingChunkDownloadEvents()
+        {
```

Comment:
Will cherry pick these changes out and put in a different PR for future review and testing.

https://github.com/Azure/azure-sdk-for-net/issues/32854